scripts/remove_after_use/register_preprint_dois_with_crossref.py

```diff
@@ -0,0 +1,90 @@
+from website.app import setup_django
+setup_django()
+
+import sys
+import time
+import logging
+
+from django.core.paginator import Paginator
+
+from osf.models import PreprintService
+from osf_tests.factories import ProjectFactory, PreprintFactory, AuthUserFactory, PreprintProviderFactory
+
+from website.identifiers.clients import CrossRefClient
+
+from website.app import init_app
+from website import settings
+
+
+logger = logging.getLogger(__name__)
+
+PAGE_SIZE = 5
+
+
+def register_existing_preprints_with_crossref(dry=True):
+    """This should do three things:
+    one, create a CSV of all preprints to be migrated for use by CNRI,
+    two, send metadata about the new preprint DOIs to CrossRef for registration,
+    three, add the new crossref specific DOIs to the preprint with the category 'doi' while marking the
+    legacy_doi identifier as deleted.
+    """
+    paginator = Paginator(
+        PreprintService.objects.filter(identifiers__category='legacy_doi', identifiers__deleted__isnull=True).order_by('pk'),
+        PAGE_SIZE
+    )
+    client = CrossRefClient(base_url=settings.CROSSREF_URL)
+
+    # Write the header for the "good" dois to convert
+    with open('preprint_doi_conversions.csv', 'w') as doi_file:
+        doi_file.write('legacy_doi, new_doi\n')
+
+    # Keep track of any faulty DOIs that won't be converted for now
+    with open('preprints_with_faulty_dois.csv','w') as bad_dois:
+        bad_dois.write('guid, legacy_doi, new_doi\n')
+
+    for page_num in paginator.page_range:
+        page = paginator.page(page_num)
+        preprints_to_migrate = []
+        batch_metadata = []
+        for preprint in page:
+            legacy_doi_identifier = preprint.get_identifier(category = 'legacy_doi')
+            new_doi = client.build_doi(preprint)
+
+            # Some DOIs it seems have None as their handle -- keep track of those and don't port them for now
+            if 'None' in legacy_doi_identifier.value:
+                with open('preprints_with_faulty_dois.csv','a') as bad_dois:
```

Comment:
To avoid the possibility of an I/O error, what if you open the file descriptor just once, before the loop?

Even better: generate the CSV and attach it to an email to send to support@osf.io . That way, we wouldn't have to transfer the files from the prod container to local, then transfer from local to Sarah. There's an example of doing this: https://github.com/CenterForOpenScience/osf.io/blob/f536db8d2872c423e98be161d145f7958072298d/scripts/generate_prereg_csv.py

Reply:
Good call on the email for the CSV! Will absolutely do that.

Also, I do open the CSV first before the loop to write the headers, right [hrr right thrr](https://github.com/CenterForOpenScience/osf.io/pull/8344/files#diff-429703a38e1b3332e43c2621737691d0R38), is something like that what you meant?

Reply:
Currently, the CSV files get opened multiple times--once to write the headers, then a bunch of times in the loop. Better to open the files just once.

But if you go with the email approach, you don't even need to open a file. Just create a BytesIO object and write the csv in-memory. You can use the `csv` module to ensure proper formatting (see the linked example).